### PR TITLE
Detangle attach detach from  gce

### PR DESCRIPTION
**Pull request by ghost**
This detangles attach and detach operations from the GCE PD plugin and implements the new Attacher/Detacher interface.

@markturansky @saad-ali What do you guys think ?

---

This change is [Reviewable](http://reviewable.k8s.io/reviews/kubernetes/kubernetes/21709)

### CODE REVIEW

**Comment by saad-ali** [comment]
This is a good start. We need to untangle the attach/mount and unmount/detach code for the https://github.com/kubernetes/kubernetes/issues/21931 redesign.

But it needs to be even more significant: we need to break "attach" into "device attach" and "verify device attached". And break "mount" into "mount device" and "bind mount pod".

So, we should define the following volume-specific hooks:
* `AttachDevice`
  * Attach specified device to specified node.
* `WaitForDeviceToAttach`
  * Examine local machine to determine if the specified volume has been successfully attached to it.
  * Wait some period of time and recheck, if it has not.
  * Fail if the check does not succeed within a reasonable amount of time.
* `DetachDevice`
  * Detach specified device from specified node.
* `WaitForDeviceToDetach`
  * Examine local machine to determine if the specified volume has been successfully detached from it.
  * Wait some period of time and recheck, if it has not.
  * Fail if the check does not succeed within a reasonable amount of time.
* `MountDevice`
  * Mount specified volume to global mount path (i.e. `/var/lib/kubelet/plugins/kubernetes.io/{plugin}/mounts/{uniqueVolumeName}/`).
* `UnmountDevice`
  * Unmount specified device from global mount path.

The bind mount/unmount code should be common, so we can stick them in a util for now.
* `BindMount`
  * Bind mounts the specified volume global mount path to the pod specific mount location (i.e. `/var/lib/kubelet/pods/{podID}/volumes/{sanatizedPluginName}/{podSpecVolumeName}/`).
* `BindUnmount`
  * Unmount bind mount between the specified volume global mount path and the pod specific mount location`

After the refactor, the `Setup` method for each plugin should call `AttachDevice`, `WaitForDeviceToAttach`,`MountDevice` or the common `BindMount`. And the `TearDown` method should call `DetachDevice`, `WaitForDeviceToDetach`, and `UnmountDevice` or the common `BindUnmount`.

Let's do this for the GCE plugin in this PR, and, then like you said, it'll be easier to follow the same pattern for the other plugins.

**Comment by swagiaal** [comment]
@saad-ali it sounds like we'll need to change the Attacher/Detacher interfaces to reflect the above. I'll do that in a separate PR first then rebase this on that. WDYT ?

**Comment by saad-ali** [comment]
> @saad-ali it sounds like we'll need to change the Attacher/Detacher interfaces to reflect the above. I'll do that in a separate PR first then rebase this on that. WDYT ?

Yes, that makes sense. @markturansky had a giant PR https://github.com/kubernetes/kubernetes/pull/19599 renaming builder/cleaner to mounter/unmounter we need to get that in too.

**Comment by swagiaal** [comment]
@saad-ali I am having second thoughts about the modularization suggested above. While it makes sense in terms of common functionality it does not make sense to expand the Attacher/Detacher interface to encompass things like `WaitForDeviceToAttach` `MountDevice` as these will have to run on the kubelet and cannot be fun from the master. So the objects implementing Attacher/Detacher will not be execution location independent.

To do this we would need:
```go
type Attacher interface {
  Attach()
}

type KubeletAttacher interface {
  WaitForDeviceToAttach()
  MountDevice()
}
```
WDYT? Do you want to proceed that way or leave this PR as is ?

**Comment by saad-ali** [comment]
> it does not make sense to expand the Attacher/Detacher interface to encompass things like WaitForDeviceToAttach MountDevice as these will have to run on the kubelet and cannot be fun from the master

Agreed; I did not mean to imply that the attacher should home to all these methods. I want the plugins broken apart so that they define each of these methods for the sake of clarity and code reuse. But, what interface each method is defined in needs to be thought through. I agree that only the `Attach()` method makes sense in the `Attacher` interface. Both `WaitForDeviceToAttach()` and `MountDevice()` should be in a separate interface. Not a fan of the name `KubeletAttacher` but I like the logical grouping.

**Comment by swagiaal** [comment]
@saad-ali updated and based this on the new interfaces PTAL

File: pkg/volume/volume.go [hunk 1]
```diff
@@ -145,28 +144,28 @@ type Attacher interface {
...
-	MountDevice(devicePath string, deviceMountPath string, mounter mount.Interface) error
+	MountDevice(spec *Spec, devicePath string, deviceMountPath string, mounter mount.Interface) error
 }
 
 // Detacher can detach a volume from a node.
 type Detacher interface {
 
 	// Detach the given volume from the given host.
-	Detach(deviceMountPath string, hostName string) error
+	Detach(cloudProvider cloudprovider.Interface, deviceMountPath string, hostName string) error
```
Comment (saad-ali) [inline pkg/volume/volume.go]: Not all volume plugins will have a cloud provider. So we should not require one in the Detach interface. Plugins that require one can get it through the VolumeHost.

Comment (saad-ali) [inline pkg/volume/volume.go]: Ditto with Attach above.

File: pkg/volume/gce_pd/attacher.go [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+func (plugin *gcePersistentDiskPlugin) NewAttacher() (volume.Attacher, error) {
+	return &gcePersistentDiskAttacher{}, nil
+}
+
+func (attacher *gcePersistentDiskAttacher) Attach(cloudProvider cloudprovider.Interface, spec *volume.Spec, hostName string) error {
+	volumeSource, readOnly := getVolumeSource(spec)
+	pdName := volumeSource.PDName
+
+	// Block execution until any pending detach operations for this PD have completed
+	attachDetachMutex.LockKey(pdName)
```
Comment (saad-ali) [inline pkg/volume/gce_pd/attacher.go]: As soon as your volume manager is in and mount/unmount attach/detach are no longer happening async we can get rid of this hack.

File: pkg/volume/gce_pd/attacher.go [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+	defer attachDetachMutex.UnlockKey(pdName)
+
+	gceCloud, err := getCloudProvider(cloudProvider)
+	if err != nil {
+		return err
+	}
+
+	for numRetries := 0; numRetries < maxRetries; numRetries++ {
```
Comment (saad-ali) [inline pkg/volume/gce_pd/attacher.go]: Once your volume manager is in (and my attach/detach controller), it can be responsible for retrying so we'll be able to simplify this code as well then.

File: pkg/volume/gce_pd/attacher.go [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+		var attached bool
+		attached, err = gceCloud.DiskIsAttached(pdName, hostName)
```
Comment (saad-ali) [inline pkg/volume/gce_pd/attacher.go]: We can't do this (succeed if volume is already) yet. If the disk is attached and you succeed here, since the volume attach/mount and detach/unmount loops are still async a bunch of race conditions that are hidden will manifest them selves.

We can do this only after your volume manager is in--then we don't have to worry about rapid pod creation deletion creating two attaches before a detach.

By the way the test/e2e/pd.go are not run as the default set on PRs, so please run them manually to confirm they are passing (they are part of the slow suite, so after PR is merged they could break and block the submit queue). A good way to catch some of the races is to run theses tests back to back for 30min to 1hr.

File: pkg/volume/gce_pd/attacher.go (hunk 3 above)
Comment (swagiaal) [inline pkg/volume/gce_pd/attacher.go]: So if we remove the `DiskIsAttached` check as you suggested below, there will no longer be a predicate upon which to base the retry. We will have to rely on the sync loop (actually in this case the pod worker thread) to do the provide the retry. WDYT ?

File: pkg/volume/gce_pd/attacher.go (hunk 4 above)
Comment (swagiaal) [inline pkg/volume/gce_pd/attacher.go]: > We can't do this (succeed if volume is already) yet. If the disk is attached and you succeed here, since the volume attach/mount and detach/unmount loops are still async a bunch of race conditions that are hidden will manifest them selves.

Oh yes; good catch.. I forgot about that scenario :/



> By the way the test/e2e/pd.go are not run as the default set on PRs, so please run them manually to confirm they are passing (they are part of the slow suite, so after PR is merged they could break and block the submit queue).

Yup.. I have been running that test regularly as part of this work.



> A good way to catch some of the races is to run theses tests back to back for 30min to 1hr.

That I have not done.. will do

File: pkg/volume/gce_pd/attacher.go (hunk 3 above)
Comment (swagiaal) [inline pkg/volume/gce_pd/attacher.go]: Actually the update PR retries on failure of the Attach/Detach call to the cloud provider. I personally like leaving retries to the sync loop but I hesitate to destabilize this code too much in this change which is already pretty big.

**Comment by swagiaal** [comment]
@saad-ali PTAL.
I have run the pd test and it passed once.. currently running on a continuous loop.
I will post an update if I see any flakyness

**Comment by swagiaal** [comment]
@k8s-bot test this. #IGNORE

File: pkg/volume/gce_pd/attacher.go [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+type gcePersistentDiskAttacher struct{}
+
+var _ volume.Attacher = &gcePersistentDiskAttacher{}
+
+var _ volume.AttachableVolumePlugin = &gcePersistentDiskPlugin{}
+
+// Singleton key mutex for keeping attach/detach operations for the
+// same PD atomic
+// TODO(swagiaal): Once the Mount/Unmount manager is implemented this
+// will no longer be needed and should be rmeoved
```
Comment (saad-ali) [inline pkg/volume/gce_pd/attacher.go]: nit typo: removed

File: pkg/volume/gce_pd/attacher.go (hunk 3 above)
Comment (saad-ali) [inline pkg/volume/gce_pd/attacher.go]: > So if we remove the DiskIsAttached check as you suggested below, there will no longer be a predicate upon which to base the retry

Retry would be based on the AttachDisk call failing immediately with err (unlikely).

> I personally like leaving retries to the sync loop but I hesitate to destabilize this code too much in this change which is already pretty big.

Yes, it would be a bad idea to remove the retries and leave them to the sync loop for now. But once the volume manager exists, then it is safe to let the volume manager do the retries.

**Comment by saad-ali** [comment]
LGTM pending rebase on changes to underlying PR.

Looks like the following E2E test is failing:
`[Fail] [k8s.io] Dynamic provisioning [k8s.io] DynamicProvisioner [It] should create and delete persistent volumes `
with:
`May  9 09:14:43.003: INFO: At {2016-05-09 08:59:42 -0700 PDT} - event for pvc-volume-tester-v4h5m: {kubelet e2e-gce-builder-3-3-minion-ohnk} FailedSync: Error syncing pod, skipping: Mount failed: exit status 32`

**Comment by swagiaal** [comment]
> Looks like the following E2E test is failing:
> [Fail] [k8s.io] Dynamic provisioning [k8s.io] DynamicProvisioner [It] should create and delete persistent volumes
> with:
> May 9 09:14:43.003: INFO: At {2016-05-09 08:59:42 -0700 PDT} - event for pvc-volume-tester-v4h5m: {kubelet e2e-gce-builder-3-3-minion-ohnk} FailedSync: Error syncing pod, skipping: Mount failed: exit status 32

yup.. looking into that currently

File: pkg/kubelet/volumes.go [hunk 6]
```diff
@@ -143,7 +143,8 @@ func (kl *Kubelet) mountExternalVolumes(pod *api.Pod) (kubecontainer.VolumeMap,
 			return nil, err
 		}
 		if attacher != nil {
-			err = attacher.Attach(volSpec, kl.hostname)
+			host := &volumeHost{kl}
+			err = attacher.Attach(host, volSpec, kl.hostname)
```
Comment (rootfs) [inline pkg/kubelet/volumes.go]: aws and gce `AttachDisk` return different results: aws returns device path and error, while gce just returns error. It will be great if `attacher.Attach` returns the union of both cases. Otherwise, the aws has to remember the attached device path.

Same to `DetachDisk`

Comment (saad-ali) [inline pkg/kubelet/volumes.go]: Attach/Detach should only return error. The attach/detach controller, for example, doesn't care where the device is mounted. Instead the `WaitForAttach`/`WaitForDetach` methods should return that info.

File: pkg/volume/gce_pd/gce_pd.go [hunk 7]
```diff
@@ -317,35 +295,18 @@ func (c *gcePersistentDiskUnmounter) TearDownAt(dir string) error {
...
 		return err
 	}
 	if notMnt {
-		if err := os.Remove(dir); err != nil {
-			return err
-		}
+		return os.Remove(dir)
 	}
-	return nil
+	return fmt.Errorf("Failed to unmount volume dir")
```
Comment (rootfs) [inline pkg/volume/gce_pd/gce_pd.go]: why create and return err?

Comment (rootfs) [inline pkg/volume/gce_pd/gce_pd.go]: i see, you changed notMnt.

File: pkg/kubelet/volumes.go (hunk 6 above)
Comment (swagiaal) [inline pkg/kubelet/volumes.go]: I think the challenge there is that the path of the device is determined by the call to AttachDisk not from the disk id as is the case with GCE. @rootfs is there a way around this ? I would say open a PR with the work done for AWS EBS and base that on another PR which has all the changes to the Attacher/Detacher interface that you need and let's carry on this discussion there. Otherwise we are trying to hit a moving target.

**Comment by swagiaal** [comment]
@saad-ali this update contains a few changes:
- Rebase on the updated Attacher interface removing the requirement to pass host and storing that in the attacher.
- Rebase on #25502 and store the volume spec in the Attacher which adds support for PVC volumes (this should fix the failing test case)
- Add a check for the global mount in the Attach call before calling to the cloud provider. Things had slowed down considerably without that check.
I thing all changes are in the top commit.

P.S. PD e2e test run still pending.. my cluster is taking a while to set up :/

File: pkg/kubelet/volumes.go (hunk 6 above)
Comment (saad-ali) [inline pkg/kubelet/volumes.go]: > I think the challenge there is that the path of the device is determined by the call to AttachDisk not from the disk id as is the case with GCE

The `waitForAttach` call for the AWS EBS plugin will have to issue another call to AWS to figure out if the volume is attached, and if so where it is mounted. AWS [`describe-volumes`](http://docs.aws.amazon.com/cli/latest/reference/ec2/describe-volumes.html), for example, returns the state (e.g. `"State": "attached"`) and the mount path (e.g. `"Device": "/dev/sda1"`.

> I would say open a PR with the work done for AWS EBS and base that on another PR which has all the changes to the Attacher/Detacher interface that you need and let's carry on this discussion there. Otherwise we are trying to hit a moving target.

Agreed, further discussion here is unwarranted. We can discuss on the AWS EBS PR. I'd argue that no further changes (other than https://github.com/kubernetes/kubernetes/pull/25523) are needed to the Attacher/Detacher interface.

File: pkg/controller/persistentvolume/persistentvolume_provisioner_controller.go [hunk 8]
```diff
@@ -506,6 +506,10 @@ func (c *PersistentVolumeProvisionerController) NewWrapperMounter(volName string
 	return nil, fmt.Errorf("NewWrapperMounter not supported by PVClaimBinder's VolumeHost implementation")
 }
 
+func (c *PersistentVolumeProvisionerController) NewWrapperAttacher(volName string, spec volume.Spec, pod *api.Pod) (volume.Attacher, error) {
```
Comment (saad-ali) [inline pkg/controller/persistentvolume/persistentvolume_provisioner_controller.go]: Looks like the wrapper is from https://github.com/kubernetes/kubernetes/pull/25502. We can address comments on it there.

**Comment by saad-ali** [comment]
I think I'm largely good with this PR alone, I have major beef with the underlying PR https://github.com/kubernetes/kubernetes/pull/25502

File: pkg/volume/gce_pd/attacher.go [hunk 9]
```diff
@@ -0,0 +1,233 @@
...
+	defer timer.Stop()
+
+	volumeSource, _ := getVolumeSource(attacher.spec)
+	pdName := volumeSource.PDName
+	partition := ""
+	if volumeSource.Partition != 0 {
+		partition = strconv.Itoa(int(volumeSource.Partition))
+	}
+
+	sdBefore, err := filepath.Glob(diskSDPattern)
```
Comment (rootfs) [inline pkg/volume/gce_pd/attacher.go]: how are we sure the disk is not already attached by now? If already attached, `sdBefore` then contains the newly attached device.

**Comment by swagiaal** [comment]
@saad-ali PTAL

File: pkg/volume/gce_pd/gce_pd.go [hunk 10]
```diff
@@ -76,25 +75,30 @@ func (plugin *gcePersistentDiskPlugin) NewMounter(spec *volume.Spec, pod *api.Po
...
-	var gce *api.GCEPersistentDiskVolumeSource
 	if spec.Volume != nil && spec.Volume.GCEPersistentDisk != nil {
-		gce = spec.Volume.GCEPersistentDisk
-		readOnly = gce.ReadOnly
+		volumeSource = spec.Volume.GCEPersistentDisk
+		readOnly = volumeSource.ReadOnly
 	} else {
-		gce = spec.PersistentVolume.Spec.GCEPersistentDisk
+		volumeSource = spec.PersistentVolume.Spec.GCEPersistentDisk
 		readOnly = spec.ReadOnly
```
Comment (saad-ali) [inline pkg/volume/gce_pd/gce_pd.go]: Not your code. But I'm curious, why not `readOnly = volumeSource.ReadOnly` here?

File: pkg/kubelet/volumes.go [hunk 11]
```diff
@@ -137,33 +155,41 @@ func (kl *Kubelet) mountExternalVolumes(pod *api.Pod) (kubecontainer.VolumeMap,
...
+			deviceMountPath := attacher.GetDeviceMountPath(volSpec)
+			notMountPoint, err := kl.mounter.IsLikelyNotMountPoint(deviceMountPath)
+			if err != nil && !os.IsNotExist(err) {
 				return nil, err
 			}
-
-			deviceMountPath := attacher.GetDeviceMountPath(&volumeHost{kl}, volSpec)
-			if err = attacher.MountDevice(volSpec, devicePath, deviceMountPath, kl.mounter); err != nil {
-				return nil, err
+			if notMountPoint {
```
Comment (saad-ali) [inline pkg/kubelet/volumes.go]: Again not your code, but this `IsLikelyNotMountPoint` stuff always confuses me. Over here we decide to attach and mount if the path is not a mount point? When would the opposite case happen?

File: pkg/volume/gce_pd/attacher.go (hunk 5 above)
Comment (saad-ali) [inline pkg/volume/gce_pd/attacher.go]: poke :)

File: pkg/volume/gce_pd/attacher.go (hunk 9 above)
Comment (saad-ali) [inline pkg/volume/gce_pd/attacher.go]: If the device is found this hack doesn't matter anyway. It was put in place to ensure the device shows up. I'm pretty sure we can get rid of the hack now. But maybe as a separate deliberate action in another PR, not here.

**Comment by saad-ali** [comment]
No more beef :)

Mostly looks good to me. Couple of questions, and some unaddressed nits. Also, could you verify the GCE PD tests are stable (run back to back for a while).

File: pkg/volume/gce_pd/gce_pd.go (hunk 10 above)
Comment (swagiaal) [inline pkg/volume/gce_pd/gce_pd.go]: it is carried over from the pvc specification in the pod definition. The spec here represents the PV which does not know whether readOnly was specified in the pod.

File: pkg/kubelet/volumes.go (hunk 11 above)
Comment (swagiaal) [inline pkg/kubelet/volumes.go]: This function keeps getting called every iteration of the  syncloop (and more) so the SetUp functions used the existence of a global which which is mounting _something_ to indication that their work is done and short circuit calling to the cloud provider in every iteration of the loop. During the first iteration the global path either does not exist or is not a mount point.
Without this check pods stay in containerCreating for over a minute and a half

File: pkg/volume/gce_pd/gce_pd.go (hunk 10 above)
Comment (saad-ali) [inline pkg/volume/gce_pd/gce_pd.go]: Ah right because the `PersistentVolumeClaimVolumeSource` defines a `ReadOnly` field that takes precedence. That's pretty confusing, we ignore the `ReadOnly` field on the PV object and let the `ReadOnly` field on the PVC object take precedence. Might be worth revisiting this logic in the future. Thanks.

File: pkg/kubelet/volumes.go (hunk 11 above)
Comment (saad-ali) [inline pkg/kubelet/volumes.go]: Ok, that makes sense for GCE. What about volumes that don't do global mount, like NFS? Also, why is `mounter.SetUp(fsGroup)` called out side of this. Is it because we expect plugins that do Attach to all do a global mount?

Comment (swagiaal) [inline pkg/kubelet/volumes.go]: presumably all volumes which implement Attacher (ie `attacher != nil`) do do the global mount. I think that assumption is fair since we expect them to implement `GetDeviceMountPath`
So NFS wont be handled inside this ` if attach != nil` statement.
`mounter.SetUp()` is expected to implement its own short circuiting and it is implemented by both Attachable and non-Attachable plugins

**Comment by swagiaal** [comment]
> Also, could you verify the GCE PD tests are stable (run back to back for a while).

Done.. PD test looks stable

File: pkg/kubelet/volumes.go (hunk 11 above)
Comment (saad-ali) [inline pkg/kubelet/volumes.go]: Ack. I still don't like having Attacher and Mounter be two different "types" of plugins. I'd love to see, all plugins that are attachers also implement the mounter interface but we can do that refactor at another point.

**Comment by saad-ali** [comment]
LGTM

**Comment by pmorie** [comment]
@k8s-bot test this issue #25812

**Comment by swagiaal** [comment]
readding LGTM after rebase...

File: pkg/volume/gce_pd/attacher.go [hunk 12]
```diff
@@ -0,0 +1,220 @@
...
+			continue
+		}
+
+		return nil
+	}
+
+	return err
+}
+
+func (attacher *gcePersistentDiskAttacher) WaitForAttach(spec *volume.Spec, timeout time.Duration) (string, error) {
```
Comment (jsafrane) [inline pkg/volume/gce_pd/attacher.go]: I know I am too late and `Attacher` interface is already merged, but `Attach()` and `WaitForAttach()` should not be part of the same interface - one will be implemented by kube-controller-manager and requires `VolumeHost.GetCloudProvider()` and the second one by kubelet and should **not** require cloud instance - one reason why we do the controller is not to have AWS credentials on the node.

Comment (saad-ali) [inline pkg/volume/gce_pd/attacher.go]: I agree. But for slightly different reasons. Both could require cloud provider--attach for attaching, wait for attach because some plugins like AWS EBS, waitForAttach would need to first call the cloud provider to figure out where on the node to expect the volume (it is not deterministic). But attach would never require the kubelet specific volume host stuff where as waitForAttach would.

Comment (swagiaal) [inline pkg/volume/gce_pd/attacher.go]: @jsafrane also if you follow the early discussion on #24557 we first tried to create an Attacher interface and a PhysicalAttacher interface where Attacher would run in the master and PhysicalAttacher runs in the kubelet, but the number of interfaces was getting out of hand so we decided to tackle this by storing minimal information int the Attacher object and instead passing everything that is needed through arguments to individual functions. Functions which require cloud stuff would only be run on master and functions which look at the local file system would only run on kubelet

**Comment by swagiaal** [comment]
Readding LGTM after rebase

**Comment by saad-ali** [comment]
Going to revert this PR since it appears to have destabilized pd E2E tests: https://github.com/kubernetes/kubernetes/issues/26076
